spectator/http.py

```diff
@@ -65,28 +71,35 @@ def post_json(self, uri, data):
 
         logger.debug("posting data to %s, payload: %s", uri, entity)
         request = urllib2.Request(uri, self._compress(entity), headers)
-
         start = self._registry.clock().monotonic_time()
-        try:
-            response = urllib2.urlopen(request, timeout=self._timeout)
-            self._add_status_tags(tags, response.code)
-            msg = self._read_response(response)
-            logger.debug("request succeeded (%d): %s", response.code, msg)
-        except urllib2.HTTPError as e:
-            self._add_status_tags(tags, e.code)
-            msg = self._read_error(e)
-            logger.warning("request failed (%d): %s", e.code, msg)
-        except urllib2.URLError as e:
-            error = type(e).__name__
-            tags["status"] = error
-            tags["statusCode"] = error
-            logger.warning("request failed: %s", e)
-        except:
-            e = sys.exc_info()[0]
-            error = type(e).__name__
-            tags["status"] = error
-            tags["statusCode"] = error
-            logger.warning("request failed: %s", e)
+
+        attempt = 1
+
+        while attempt <= max_attempts:
+            try:
+                response = urllib2.urlopen(request, timeout=self._timeout)
+                self._add_status_tags(tags, response.code)
+                msg = self._read_response(response)
+                logger.debug("request succeeded, code=%d attempt=%d/%d: %s", response.code, attempt, max_attempts, msg)
+                attempt = max_attempts + 1
+            except urllib2.HTTPError as e:
+                self._add_status_tags(tags, e.code)
+                msg = self._read_error(e)
+                logger.debug("request failed, code=%d attempt=%d/%d: %s", e.code, attempt, max_attempts, msg)
+                time.sleep(retry_delay)
+                if e.code == 429 or int(e.code / 100) == 500:
```

Comment:
`int(e.code / 100) == 500`, if you divide by 100 shouldn't it be 5 rather than 500? I think `e.code >= 500` would also be sufficient as there is no 6xx or higher.

Reply:
Ah, yes. I should include a test case to post a 503 so that we can at least see the log output.